### PR TITLE
Add "runtime_params" resolver to allow overriding of config with CLI params

### LINKED ISSUE #2531: Enable overriding configuration via variable interpolation for all configurations other than just `parameters` for OmegaConfigLoader
## Description
Is your feature request related to a problem? A clear and concise description of what the problem is: "I'm always frustrated when ..."

With #2467 we resolves OCL earlier and parameters propogate correctly for `parameters`. We cannot do the same for `catalog` or other configuration yet due to the problem of "residual" configurations. The problem is specific to `OmegaConfigLoader`, because it requires merging the dictionaries before resolving.

Take this simple example

```yaml
# catalog.yml
my_dataset:
   path: ${base_path}/my_file.txt
```

## Expected Result
`kedro run --params base_path:my_bucket` this should work and replace the value correctly.

## Current Result
The value will not propagate for configurations other than `parameters`. This is because we need to merge `aggregated_configs` with `runtime_params`, which produce invalid catalog entries.

The compiled catalog may look like this
```yaml
# catalog.yml
my_dataset:
  path: ${base_path}/my_file.txt


base_path: my_bucket # <-- This is not a dataset, ideally we need to remove it or bypass it.
```

## Context
Why is this change important to you? How would you use it? How can it benefit other users?
This is essential to enable common use case like providing a bucket name for `catalog` via CLI.
`kedro run --params base_path:my_bucket` , with TemplatedConfigLoader it doesn't work either, but there are [workarounds](https://github.com/kedro-org/kedro/discussions/1782). In contrast, there is no workaround for OmegaConfigLoader.

## Possible Implementation
(Optional) Suggest an idea for implementing the addition or change.
1. Figure out a way to remove the residual configurations

The implementation should make sure above case will work correctly. It should also work for arbitrary configurations, i.e. `config_loader["spark"] or `config_loader["my_random_config"]` as long as it's registered with `config_patterns`

## Possible Alternatives
(Optional) Describe any alternative solutions or features you've considered.

> How does “global” fit in the picture? It will be clearer when we finished Change 1
> Antony thinks that with the validation change, it shouldn’t be matter even when the “residuals” are left in the config, Python config also don’t do anything special to extra configurations.
> Ivan think this doesn’t solve the problem in a generic way, config_lodaer[“global”] is purposed, but this doesn’t solve all the problem because OmegaConfigLoader requires to merge dictionary before interpolating. This shouldn’t be a limitation, but will requires more work to implement a solution
> Merel has some concern about the messaging of “meta_params”, will people really understand which argument to be used.
> kedro run --params=base_path:s3_bucket will not work for catalog yet, but agree that it would be nice to support.

### PR DESCRIPTION
**Pull request by ankatiyar**
## Description

Resolve #2531 

### `runtime_params` resolver
- similar to globals - registered using `_registed_runtime_params_resolver()`
- Uses `_get_runtime_value()` to get the interpolated value, also similar to globals' `_get_globals_value()`
### Changes to `globals`
- Read globals and populate `self._globals` in the `__init__` function. 
- When global values are read by `_get_globals_value()` it uses the `self._globals` dict. 
- This is so when globals are first read, the `runtime_params` resolver is ***cleared*** and then ***registered again*** in the `__getitem__` fn
- Also helps that we don't need to read `self["globals"]` each time to resolve wherever globals is used.
### On the topic of nested resolvers
***Need feedback from reviewers on this***
- I disable the use of `runtime_params` in `globals.yml` by disabling the `runtime_params` in `__getitem__` if the key is "globals" and then registering it again. Throws an `UnsupportedInterpolationType` error when `runtime_params:` is used in globals.
- Disabling globals within `runtime_params` is tricky, I haven't found a clean way to do it.

However, I think allowing globals to be used as a default value for parameters with `runtime_params` resolver is valid and might be useful and we probably should allow this - 
```yaml
my_param : "${runtime_params: y, ${globals: x, 34}}"
```

### TODO

- [x] Test
- [x] Docs
- [x] Release Notes

## Checklist

- [x] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [x] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [x] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [x] Added tests to cover my changes
- [x] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

File: kedro/config/omegaconf_config.py [hunk 1]
```diff
@@ -178,9 +186,15 @@ def __getitem__(self, key) -> dict[str, Any]:
...
-            base_path, patterns, key, processed_files, read_environment_variables
-        )
+        try:
+            base_config = self.load_and_merge_dir_config(
+                base_path, patterns, key, processed_files, read_environment_variables
+            )
+        except UnsupportedInterpolationType:
+            raise UnsupportedInterpolationType(
+                "The `runtime_params:` resolver is not supported for globals."
+            )
```
Comment (merelcht) [inline kedro/config/omegaconf_config.py:196]: This could potentially be raised because of something else right? So we should make sure that this specific message only shows when the user is trying to use `runtime_params` with globals.

**Review by merelcht — COMMENTED** [review]
This looks great @ankatiyar ! I have also tried some simple use cases and it works perfectly 👍  I agree that it's should be allowed for users to use `globals` as a default inside the `runtime_params` resolver. But definitely the other way around doesn't make sense. 

Are you adding docs as part of this PR or separately?

File: kedro/config/omegaconf_config.py (hunk 1 above)
Comment (ankatiyar) [reply kedro/config/omegaconf_config.py]: I've updated the condition to only show this message if `runtime_params` is used.

File: RELEASE.md [hunk 2]
```diff
@@ -11,6 +11,7 @@
 
 ## Major features and improvements
 * Allowed using of custom cookiecutter templates for creating pipelines with `--template` flag for `kedro pipeline create` or via `template/pipeline` folder.
+* Allowed overriding of configuration keys with runtime parameters using the `runtime_params` resolver.
```
Comment (merelcht) [inline RELEASE.md:14]: It would be good to still specify this is for the `OmegaConfigLoader` since the other two config loaders are still around.

**Review by merelcht — APPROVED** [review]
This is great! I think the docs are very clear too 👍

File: docs/source/configuration/advanced_configuration.md [hunk 3]
```diff
@@ -300,9 +296,31 @@ You can also provide a default value to be used in case the global variable does
 ```yaml
 my_param: "${globals: nonexistent_global, 23}"
 ```
-If there are duplicate keys in the globals files in your base and run time environments, the values in the run time environment
+If there are duplicate keys in the globals files in your base and runtime environments, the values in the runtime environment
 will overwrite the values in your base environment.
```
Comment (stichbury) [inline docs/source/configuration/advanced_configuration.md:300]:
```suggestion
overwrite the values in your base environment.
```

File: docs/source/configuration/advanced_configuration.md [hunk 4]
```diff
@@ -300,9 +296,31 @@ You can also provide a default value to be used in case the global variable does
...
 
+### How to override configuration with runtime parameters with the `OmegaConfigLoader`
+
+Kedro allows you to [specify runtime parameters for the `kedro run` command with the `--params` CLI option](parameters.md#how-to-specify-parameters-at-runtime). These runtime parameters
+are added to the `KedroContext` and merged with parameters from the configuration files to be used in your project's pipelines and nodes. From Kedro `0.18.14`, you can use the
+`runtime_params` resolver to indicate that you want to override values of certain keys in your configuration with runtime parameters provided through the CLI option.
+This resolver can be used across different configuration types, such as parameters, catalog, and more, except for "globals".
+Consider this `catalog.yml` file :
```
Comment (stichbury) [inline docs/source/configuration/advanced_configuration.md:308]:
```suggestion

Consider this `catalog.yml` file:
```

**Review by stichbury — APPROVED** [review]
This looks good to me, thanks @ankatiyar!

I made a commit to the branch to move the table of contents that links to the "How to " subsections to the top of the page (as there was some repetition with it in the place previously sited).

File: RELEASE.md (hunk 2 above)
Comment (ankatiyar) [inline RELEASE.md:14]:
```suggestion
* Allowed overriding of configuration keys with runtime parameters using the `runtime_params` resolver with `OmegaConfigLoader`.
```

File: docs/source/configuration/advanced_configuration.md [hunk 5]
```diff
@@ -300,9 +296,32 @@ You can also provide a default value to be used in case the global variable does
...
+This resolver can be used across different configuration types, such as parameters, catalog, and more, except for "globals".
+
+Consider this `catalog.yml` file:
+```yaml
+model_options:
+  random_state: "${runtime_params:random}"
+```
+This will allow you to pass a runtime parameter named `random` through the CLI to specify the value of `model_options.random_state` in your project's parameters:
+```bash
+kedro run --params="random=3"
```
Comment (noklam) [inline docs/source/configuration/advanced_configuration.md:316]:
```suggestion
kedro run --params random=3
```
Nit-pick

File: docs/source/configuration/advanced_configuration.md [hunk 6]
```diff
@@ -300,9 +296,32 @@ You can also provide a default value to be used in case the global variable does
...
+overwrite the values in your base environment.
+
+### How to override configuration with runtime parameters with the `OmegaConfigLoader`
 
+Kedro allows you to [specify runtime parameters for the `kedro run` command with the `--params` CLI option](parameters.md#how-to-specify-parameters-at-runtime). These runtime parameters
+are added to the `KedroContext` and merged with parameters from the configuration files to be used in your project's pipelines and nodes. From Kedro `0.18.14`, you can use the
+`runtime_params` resolver to indicate that you want to override values of certain keys in your configuration with runtime parameters provided through the CLI option.
+This resolver can be used across different configuration types, such as parameters, catalog, and more, except for "globals".
+
+Consider this `catalog.yml` file:
```
Comment (noklam) [inline docs/source/configuration/advanced_configuration.md:309]:
```suggestion
Consider this `parameters.yml` file:
```

File: kedro/config/omegaconf_config.py [hunk 7]
```diff
@@ -170,6 +175,9 @@ def __getitem__(self, key) -> dict[str, Any]:
             )
         patterns = [*self.config_patterns[key]]
 
+        if key == "globals":
+            # "runtime_params" resolver is not allowed in globals.
+            OmegaConf.clear_resolver("runtime_params")
```
Comment (noklam) [inline kedro/config/omegaconf_config.py:180]: Does that means running `config_loader["global"]` twice could get different results? User may use `config_loader["global"]` explicitly to perform logging etc. https://kedro-org.slack.com/archives/C03RKP2LW64/p1695380623784409

File: kedro/config/omegaconf_config.py [hunk 8]
```diff
@@ -339,6 +375,19 @@ def _get_globals_value(self, variable, default_value=_NO_VALUE):
...
+            runtime_oc, variable, default=default_value
+        )
+        if interpolated_value != _NO_VALUE:
+            return interpolated_value
+        else:
+            raise InterpolationResolutionError(
+                f"Runtime parameter '{variable}' not found and no default value provided."
+            )
+
     @staticmethod
```
Comment (noklam) [inline kedro/config/omegaconf_config.py:391]: 👍🏼

File: tests/config/test_omegaconf_config.py [hunk 9]
```diff
@@ -860,7 +860,7 @@ def test_not_hidden_config(self, tmp_path, hidden_path):
         assert not conf._is_hidden(hidden_path)
 
     def test_ignore_ipynb_checkpoints(self, tmp_path, mocker):
-        conf = OmegaConfigLoader(str(tmp_path), default_run_env=_BASE_ENV)
+        conf = OmegaConfigLoader(str(tmp_path), default_run_env="base")
```
Comment (noklam) [inline tests/config/test_omegaconf_config.py:863]: Why do you need to change it? isn't the default value `base` already?

File: kedro/config/omegaconf_config.py [hunk 10]
```diff
@@ -189,9 +206,19 @@ def __getitem__(self, key) -> dict[str, Any]:
...
+            )
+        except UnsupportedInterpolationType as exc:
+            if "runtime_params" in str(exc):
+                raise UnsupportedInterpolationType(
+                    "The `runtime_params:` resolver is not supported for globals."
+                )
+            else:
+                raise exc
+        # Re-register runtime params resolver incase it was deactivated
+        self._register_runtime_params_resolver()
```
Comment (noklam) [inline kedro/config/omegaconf_config.py:221]: I find this a bit hard to follow - I am not sure if this is registered or not. If I follow the logic, it assumes that it is always on and only turn off if key==global. Maybe this should move to the beginning instead? Then the `__init__` register_runtime_params_resolvers can be removed optionally.

In fact, I found this is inconsistent and I think we should treat this same as the `read_environment_variable` flag.

File: kedro/config/omegaconf_config.py (hunk 7 above)
Comment (ankatiyar) [reply kedro/config/omegaconf_config.py:180]: I don't think running `config_loader["globals"]` would return anything different each time, it just turns off the `runtime_params` resolver.

File: kedro/config/omegaconf_config.py (hunk 10 above)
Comment (ankatiyar) [reply kedro/config/omegaconf_config.py:221]: Since `replace=True` is set in `_register_runtime_params_resolver()`, I didn't add an if condition here. It'll re-register it anyway.

File: tests/config/test_omegaconf_config.py (hunk 9 above)
Comment (ankatiyar) [reply tests/config/test_omegaconf_config.py]: Accidentally left this in, reverted.

File: kedro/config/omegaconf_config.py (hunk 10 above)
Comment (ankatiyar) [reply kedro/config/omegaconf_config.py]: I've moved it to the beginning and removed the call from `__init__`.

**Review by merelcht — APPROVED** [review]
Btw the docs failure for the databricks was fixed in `develop` with https://docs.databricks.com/en/dev-tools/cli/authentication.html

File: kedro/config/omegaconf_config.py [hunk 11]
```diff
@@ -141,8 +141,17 @@ def __init__(  # noqa: too-many-arguments
...
+        try:
+            self._globals = self["globals"]
+        except MissingConfigException:
+            self._globals = {}
 
-    def __getitem__(self, key) -> dict[str, Any]:
+    def __setitem__(self, key, value):
+        if key == "globals":
+            self._globals = value
+        super().__setitem__(key, value)
```
Comment (noklam) [inline kedro/config/omegaconf_config.py:152]: I think we should have a comment explaining why we treat global differently.

**Review by noklam — APPROVED** [review]
Approved with minor comments. It's a great feature to add!

File: kedro/config/omegaconf_config.py (hunk 11 above)
Comment (ankatiyar) [inline kedro/config/omegaconf_config.py:152]:
```suggestion
    def __setitem__(self, key, value):
        if key == "globals":
        # Update the cached value at self._globals since it is used by the globals resolver
            self._globals = value
        super().__setitem__(key, value)
```
